test/test_inference.jl

```diff
@@ -720,4 +720,206 @@ end
+
+    @test haskey(result.predictions, :o)
+    @test typeof(result.predictions[:o]) <: NormalDistributionsFamily
+
+    # test #6 (single datavar missing)
+    @model function model_6()
+        y   = datavar(Float64) where {allow_missing = true}
+        x_0 = datavar(Float64)
+
+        x ~ Normal(mean = x_0, var = 1.0)
+        a ~ Normal(mean = x, var = 1.0)
+        b ~ Normal(mean = a, var = 1.0)
+        c ~ Normal(mean = a, var = 1.0)
+        d ~ Normal(mean = c + b, var = 1.0)
+
+        y ~ Normal(mean = d, var = 1.0)
+    end
+
+    result = inference(model = model_6(), data = (y = missing, x_0 = 1.0), initmessages = (a = vague(NormalMeanPrecision),), iterations = 10, free_energy = false)
+
+    @test haskey(result.predictions, :y)
+    @test typeof(result.predictions[:y]) <: NormalDistributionsFamily
+
+    # test #7 vmp model
+    data = (y = [1.0, -10.0, 5.0],)
+    @model function vmp_model(n)
+        x = randomvar(n + 1)
+        o = datavar(Float64) where {allow_missing = true}
+        y = datavar(Float64, n)
+
+        x_0 ~ NormalMeanPrecision(0.0, 100.0)
+        γ ~ GammaShapeRate(1.0, 1.0)
+
+        x_prev = x_0
+        for i in 1:n
+            x[i] ~ NormalMeanPrecision(x_prev, γ)
+            y[i] ~ NormalMeanPrecision(x[i], 1.0)
+            x_prev = x[i]
+        end
+        x[n + 1] ~ NormalMeanPrecision(x[n], 1.0)
+        o ~ NormalMeanPrecision(x[n + 1], 1.0)
+    end
+
+    constraints = @constraints begin
+        q(x_0, x, γ) = q(x_0, x)q(γ)
+    end
+
+    result = inference(
+        model = vmp_model(length(data[:y])),
+        data = data,
+        constraints = constraints,
+        free_energy = false,
+        initmarginals = (γ = GammaShapeRate(1.0, 1.0),),
+        iterations = 10,
+        returnvars = (γ = KeepEach(),),
+        predictvars = (o = KeepEach(),)
+    )
+
+    @test first(result.posteriors[:γ]) != last(result.posteriors[:γ])
+    @test first(result.predictions[:o]) != last(result.predictions[:o])
+
+    # test #8 non gaussian likelihood (single datavar missing)
+    dataset = [1.0, 0.0, 1.0, missing]
+    @model function coin_model(n)
+        y = datavar(Float64, n) where {allow_missing = true}
+
+        θ ~ Beta(4.0, 8.0)
+        for i in 1:n
+            y[i] ~ Bernoulli(θ)
+        end
+    end
+
+    result = inference(model = coin_model(length(dataset)), data = (y = dataset,))
+
+    @test typeof(last(result.predictions[:y])) <: Bernoulli
+
+    # test #9 allow_missing error handling
+    dataset = [1.0, 0.0, 1.0, missing]
+    @model function coin_model(n)
+        y = datavar(Float64, n)
+
+        θ ~ Beta(4.0, 8.0)
+        for i in 1:n
+            y[i] ~ Bernoulli(θ)
+        end
+    end
+
+    @test_throws ErrorException inference(model = coin_model(length(dataset)), data = (y = dataset,))
+
+    #test #10 free_energy error handling
+    @test_throws ErrorException inference(model = coin_model(length(dataset)), data = (y = dataset,), free_energy = true)
```

Comment:
Can we add one more tests for this case, with `data` which does not contain `missing`, but with explicit `predictvars`?

Reply:
added